src/screens/Forum/ForumHomeScreen.js

```diff
@@ -26,55 +31,104 @@ export default class ForumHomeScreen extends React.Component {
-        const posts = forumPosts.map((post) => {
-          const date = post.createdAt ? post.createdAt.toDate() : null;
-          const time = date ? date.toTimeString() : null;
-          const { currentUserID } = this.state;
+        const lastReferencedPost = forumPosts[forumPosts.length - 1];
 
-          return (
-            <ForumPost
-              belongsToCurrentUser={currentUserID === post.userID}
-              key={post.id}
-              /* Pass in userID  if it exists, other pass in null */
-              userID={post.userID ? post.userID : null}
-              time={time}
-              postID={post.id}
-              navigateToPostScreen={this.navigateToPostScreen}
-            >
-              {post.title}
-            </ForumPost>
-          );
-        });
-        this.setState({ posts, loading: false });
-      }, (error) => { /* Error handler */
-        this.setState({ errorMessage: error.message, loading: false });
-      });
+        this.setState({ forumPosts, lastReferencedPost, loading: false });
+      })
+      .catch((error) => this.setState({ errorMessage: error.message, loading: false }));
   }
 
   componentWillUnmount() {
     this.unsubscribeFromAuth();
-    this.unsubscribeFromFirestore();
   }
 
   navigateToPostScreen(postID, userID) {
     const { navigation } = this.props;
     navigation.navigate('ForumPost', { postID, userID });
   }
 
+  // Return whether user has scrolled to or near the end of the screen
+  reachedEnd({ layoutMeasurement, contentOffset, contentSize }) {
+    const { paddingToBottom } = this.state;
+    return layoutMeasurement.height + contentOffset.y >= contentSize.height - paddingToBottom;
+  }
+
+  // Fetch more data from firestore to load next posts
+  loadMore() {
+    const { postLimit, forumPosts, lastReferencedPost } = this.state;
+
+    firestore().collection('forum_posts')
+      .orderBy('createdAt', 'desc')
+      .startAfter(lastReferencedPost.createdAt)
+      .limit(postLimit)
+      .get()
+      .then((snapshot) => {
+        const newForumPosts = snapshot.docs.map((doc) => ({ ...doc.data(), id: doc.id }));
+        const newLastReferenced = newForumPosts[newForumPosts.length - 1];
+
+        this.setState({ // Store/append updated data for next postLimit# of posts in state
+          forumPosts: [...forumPosts, ...newForumPosts],
+          lastReferencedPost: newLastReferenced,
+          loading: false,
+        });
+      })
+      .catch((error) => this.setState({ errorMessage: error.message, loading: false }));
+  }
+
   render() {
-    const { posts, loading, errorMessage } = this.state;
+    const {
+      forumPosts, loading, loadingMore, errorMessage, currentUserID,
+    } = this.state;
     const { navigation } = this.props;
 
     return (
       <View style={styles.homeContainer}>
-        <ScrollView>
-          {loading && <ActivityIndicator /> }
+        <ScrollView
```

Comment:
I like that you were able to implement your own way of detecting when we reached the end! On one hand it works pretty well, but on the other hand, it turns out that the `FlatList` component that comes with React Native has this built in. In fact we probably should have used it from the beginning because its designed for creating lists of objects like this!

- Official doc for `FlatList`: https://reactnative.dev/docs/flatlist
- Decent tutorial for using `FlatList` with infinite scroll: https://scotch.io/tutorials/implementing-an-infinite-scroll-list-in-react-native 

I think it shouldn't be too bad to refactor this to use `FlatList` instead of `ScrollView`, maybe like this:

```
        <FlatList
          data={forumPosts.map((post) => {
            const time = post.createdAt ? post.createdAt.toDate().toTimeString() : null;
            const belongsToCurrentUser = currentUserID === item.userID;
            return {
              ...post,
              time,
              belongsToCurrentUser,
            };
          })}
          renderItem={({ item }) => (
            <ForumPost
              belongsToCurrentUser={item.belongsToCurrentUser}
              userID={item.userID}
              time={item.time}
              postID={item.id}
              navigateToPostScreen={this.navigateToPostScreen}
            >
              {item.title}
            </ForumPost>
          )}
          keyExtractor={(post) => post.id}
          onEndReached={this.loadMore}
          ListFooterComponent={(
            <View style={{ height: 40, justifyContent: 'center' }}>
              <ActivityIndicator />
            </View>
          )}
        />
```